applications/services/storage/storage_processing.c

```diff
@@ -260,9 +261,149 @@ static bool storage_process_file_eof(Storage* app, File* file) {
     return ret;
 }
 
+/*************** Sorting Dir Functions ***************/
+
+static bool storage_process_dir_rewind_internal(StorageData* storage, File* file);
+static bool storage_process_dir_read_internal(
+    StorageData* storage,
+    File* file,
+    FileInfo* fileinfo,
+    char* name,
+    const uint16_t name_length);
+
+static int storage_sorted_file_record_compare(const void* sorted_a, const void* sorted_b) {
+    SortedFileRecord* a = (SortedFileRecord*)sorted_a;
+    SortedFileRecord* b = (SortedFileRecord*)sorted_b;
+
+    if(a->info.flags & FSF_DIRECTORY && !(b->info.flags & FSF_DIRECTORY))
+        return -1;
+    else if(!(a->info.flags & FSF_DIRECTORY) && b->info.flags & FSF_DIRECTORY)
+        return 1;
+    else
+        return furi_string_cmp(a->name, b->name);
```

Comment:
Would it be better to use `furi_string_cmpi` for a case-insensitive sorting, given FAT is a case-insensitive filesystem?  [`fatsort`](https://fatsort.sourceforge.io/ ) does that by default.

I also use [the "natural order sorting" option](https://en.wikipedia.org/wiki/Natural_sort_order ) with `fatsort`, for example:
* Alphabetical
  1. `z11`
  2. `z2`
* Natural
  1. `z2`
  2. `z11`

I think it's a better result, but I'd understand if that's more than you'd want to handle in this pull request.

Reply:
Yes, case-sensitive sorting makes sense. But sorting in natural order is out of the scope of this pull request, I'll leave that for future improvements.